lib/srv/alpnproxy/forward_proxy.go

```diff
@@ -0,0 +1,305 @@
+	// DropUnwantedRequests drops the request if no receiver wants the request.
+	// If false, forward proxy sends the unwanted request to original host.
+	DropUnwantedRequests bool
+	// Log is the logger.
+	Log logrus.FieldLogger
+	// CloseContext is the close context.
+	CloseContext context.Context
+
+	// InsecureSystemProxy allows insecure system proxy when forwarding
+	// unwanted requests.
+	InsecureSystemProxy bool
+	// SystemProxyFunc is the function that determines the system proxy URL to
+	// use for a given request URL.
+	SystemProxyFunc func(reqURL *url.URL) (*url.URL, error)
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (c *ForwardProxyConfig) CheckAndSetDefaults() error {
+	if c.TunnelProtocol == "" {
+		c.TunnelProtocol = "https"
+	}
+	if c.Listener == nil {
+		return trace.BadParameter("missing listener")
+	}
+	if c.CloseContext == nil {
+		return trace.BadParameter("missing close context")
+	}
+	if c.Log == nil {
+		c.Log = logrus.WithField(trace.Component, "fwdproxy")
+	}
+	if c.SystemProxyFunc == nil {
+		c.SystemProxyFunc = httpproxy.FromEnvironment().ProxyFunc()
+	}
+	return nil
+}
+
+// ForwardProxy is a forward proxy that serves CONNECT tunnel requests.
+type ForwardProxy struct {
+	cfg ForwardProxyConfig
+}
+
+// NewForwardProxy creates a new forward proxy server.
+func NewForwardProxy(cfg ForwardProxyConfig) (*ForwardProxy, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &ForwardProxy{
+		cfg: cfg,
+	}, nil
+}
+
+// Start starts serving requests.
+func (p *ForwardProxy) Start() error {
+	err := http.Serve(p.cfg.Listener, p)
+	if err != nil && !utils.IsOKNetworkError(err) {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Close closes the forward proxy.
+func (p *ForwardProxy) Close() error {
+	if err := utils.CloseListener(p.cfg.Listener); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ServeHTTP serves HTTP requests. Implements http.Handler.
+func (p *ForwardProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	// Clients usually send plain HTTP requests directly without CONNECT tunnel
+	// when proxying HTTP requests. These requests are rejected as only
+	// requests through CONNECT tunnel are allowed.
+	if !IsConnectRequest(req) {
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	clientConn := hijackClientConnection(rw)
+	if clientConn == nil {
+		p.cfg.Log.Errorf("Failed to hijack client connection.")
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer clientConn.Close()
+
+	for _, receiver := range p.cfg.Receivers {
+		if receiver.Want(req) {
+			err := p.forwardClientToHost(clientConn, receiver.Addr().String(), req)
+			if err != nil {
+				p.cfg.Log.WithError(err).Errorf("Failed to handle forward request for %q.", req.Host)
+				p.writeHeader(clientConn, req.Proto, http.StatusInternalServerError)
+			}
+			return
+		}
+	}
+
+	p.handleUnwantedRequest(clientConn, req)
```

Comment:
What adding handling this by adding the last default receiver to `cfg.Recivers` that does the `handleUnwantedRequest` logic 
So instead of having two` forwardClientToHost` and `forwardClientToHost` the ConnectMethod handler can implemented own handler logic where a clientConn and req is injected.

Reply:
I think I got what you mean. let me see how to simplify this.